In [1]:
# Import the libraries used for data analysis and charts.
from pathlib import Path

import numpy as np
import pandas as pd
import plotly.express as px



In [2]:
# Load the job-postings dataset and clean its column names.
DATA_PATH = Path("data/Jobs_NYC_Postings_20260608.csv")

if not DATA_PATH.exists():
    DATA_PATH = Path("../data/Jobs_NYC_Postings_20260608.csv")

jobs = pd.read_csv(DATA_PATH, low_memory=False)
jobs.columns = jobs.columns.str.strip()

jobs.head()


,Job ID,Agency,Posting Type,# Of Positions,Business Title,Civil Service Title,Title Classification,Title Code No,Level,Job Category,...,Additional Information,To Apply,Hours/Shift,Work Location 1,Recruitment Contact,Residency Requirement,Posting Date,Post Until,Posting Updated,Process Date
0,780346,DEPT OF HEALTH/MENTAL HYGIENE,External,7,"Public Health Inspector, Bureau of Food Safety...",PUBLIC HEALTH SANITARIAN,Competitive-1,31215,01,"Public Safety, Inspections, & Enforcement",...,NaN,NaN,NaN,NYC - All Boroughs,NaN,New York City residency is generally required ...,05/12/2026,02-AUG-2026,05/12/2026,06/02/2026
1,780851,DEPT OF DESIGN & CONSTRUCTION,External,1,Project Manager,ASSOCIATE PROJECT MANAGER,Competitive-1,22427,01,"Engineering, Architecture, & Planning",...,NaN,NaN,NaN,30-30 Thomson Ave L I City Qns,NaN,New York City Residency is not required for th...,05/12/2026,11-JUL-2026,05/15/2026,06/02/2026
2,781780,DEPT OF HEALTH/MENTAL HYGIENE,External,1,"Scientist (Water Ecology), I - Bureau of Envir...",SCIENTIST (WATER ECOLOGY),Competitive-1,21538,01,"Public Safety, Inspections, & Enforcement",...,NaN,NaN,NaN,NYC - All Boroughs,NaN,New York City residency is generally required ...,05/26/2026,25-JUL-2026,05/26/2026,06/02/2026
3,778992,DEPT OF HEALTH/MENTAL HYGIENE,External,1,"Associate Laboratory Microbiologist, Bureau of...",ASSOCIATE LABORATORY MICROBIOL,Competitive-1,21514,02,Health,...,NaN,NaN,NaN,"455 First Ave., N.Y.",NaN,New York City residency is generally required ...,04/29/2026,28-JUN-2026,04/30/2026,06/02/2026
4,772547,FINANCIAL INFO SVCS AGENCY,Internal,1,Executive Assistant,ADM MANAGER-NON-MGRL FRM M1/M2,Competitive-1,1002C,03,"Technology, Data & Innovation",...,NaN,NaN,NaN,5 Manhattan West,NaN,New York City residency is generally required ...,05/26/2026,08-JUN-2026,05/26/2026,06/02/2026


In [13]:
# Create a reference table that explains each dataset column.
column_descriptions = {
    "Job ID": "Unique identifier assigned to the job posting.",
    "Agency": "NYC government agency hiring for the position.",
    "Posting Type": "Type of hiring or posting process used for the vacancy.",
    "# Of Positions": "Number of openings available for this posting.",
    "Business Title": "Plain-language or department-specific title used to describe the role.",
    "Civil Service Title": "Official NYC civil-service classification connected to eligibility, exams, and pay rules.",
    "Title Classification": "Classification of the position, such as competitive, non-competitive, or exempt.",
    "Title Code No": "Numeric code associated with the civil-service title.",
    "Level": "Grade, level, or rank within the job title when provided.",
    "Job Category": "Broad occupational category assigned to the position.",
    "Full-Time/Part-Time indicator": "Indicates whether the position is full-time or part-time.",
    "Career Level": "Experience or seniority level of the position.",
    "Salary Range From": "Minimum salary offered for the position.",
    "Salary Range To": "Maximum salary offered for the position.",
    "Salary Frequency": "Time period used for the salary, such as annual or hourly.",
    "Work Location": "Primary location where the employee will work.",
    "Division/Work Unit": "Department, division, bureau, or work unit hiring for the role.",
    "Job Description": "Summary of the position's responsibilities and duties.",
    "Minimum Qual Requirements": "Required education, experience, licenses, or other qualifications.",
    "Preferred Skills": "Additional skills, experience, or qualifications preferred by the employer.",
    "Additional Information": "Extra details about the position, eligibility, or hiring process.",
    "To Apply": "Instructions for submitting an application.",
    "Hours/Shift": "Expected work schedule, hours, or shift information.",
    "Work Location 1": "Additional or standardized work-location field.",
    "Recruitment Contact": "Contact information for questions about the recruitment or position.",
    "Residency Requirement": "Any requirement that the applicant live in NYC or another specified location.",
    "Posting Date": "Date the job posting was first published.",
    "Post Until": "Date through which applications or the posting are expected to remain open.",
    "Posting Updated": "Most recent date the posting information was updated.",
    "Process Date": "Date the record was processed or added to the dataset."
}

column_dictionary = pd.DataFrame({
    "Column": jobs.columns,
    "Description": [column_descriptions.get(column, "Description not provided.") for column in jobs.columns]
})

print(f"Showing all {len(jobs.columns)} columns")
with pd.option_context(
    "display.max_rows", None,
    "display.max_columns", None,
    "display.max_colwidth", None,
):
    display(column_dictionary)

Showing all 30 columns


,Column,Description
0,Job ID,Unique identifier assigned to the job posting.
1,Agency,NYC government agency hiring for the position.
2,Posting Type,Type of hiring or posting process used for the vacancy.
3,# Of Positions,Number of openings available for this posting.
4,Business Title,Plain-language or department-specific title used to describe the role.
5,Civil Service Title,"Official NYC civil-service classification connected to eligibility, exams, and pay rules."
6,Title Classification,"Classification of the position, such as competitive, non-competitive, or exempt."
7,Title Code No,Numeric code associated with the civil-service title.
8,Level,"Grade, level, or rank within the job title when provided."
9,Job Category,Broad occupational category assigned to the position.


In [17]:
# Summarize each column's type, completeness, and unique values.
profile = pd.DataFrame({
    "Column": jobs.columns,
    "Description": [column_descriptions.get(column, "Description not provided.") for column in jobs.columns],
    "Data Type": jobs.dtypes.astype(str).values,
    "Non-Null Values": jobs.notna().sum().values,
    "Missing Values": jobs.isna().sum().values,
    "Missing Percent": (jobs.isna().mean() * 100).round(1).values,
    "Unique Values": jobs.nunique(dropna=True).values,
})

print(f"Rows: {jobs.shape[0]:,}")
print(f"Columns: {jobs.shape[1]:}")

with pd.option_context(
    "display.max_rows", None,
    "display.max_columns", None,
    "display.max_colwidth", None,
):
    display(profile)

Rows: 2,362
Columns: 30


,Column,Description,Data Type,Non-Null Values,Missing Values,Missing Percent,Unique Values
0,Job ID,Unique identifier assigned to the job posting.,int64,2362,0,0.0,1228
1,Agency,NYC government agency hiring for the position.,str,2362,0,0.0,63
2,Posting Type,Type of hiring or posting process used for the vacancy.,str,2362,0,0.0,2
3,# Of Positions,Number of openings available for this posting.,int64,2362,0,0.0,29
4,Business Title,Plain-language or department-specific title used to describe the role.,str,2362,0,0.0,908
5,Civil Service Title,"Official NYC civil-service classification connected to eligibility, exams, and pay rules.",str,2362,0,0.0,287
6,Title Classification,"Classification of the position, such as competitive, non-competitive, or exempt.",str,2362,0,0.0,6
7,Title Code No,Numeric code associated with the civil-service title.,str,2362,0,0.0,306
8,Level,"Grade, level, or rank within the job title when provided.",str,2362,0,0.0,19
9,Job Category,Broad occupational category assigned to the position.,str,2362,0,0.0,103


In [22]:
# Plot the percentage of missing values in each incomplete column.
missing_cols = (
    profile.loc[profile["Missing Percent"] > 0]
    .sort_values("Missing Percent", ascending=True)
)

fig = px.bar(
    missing_cols,
    x="Missing Percent",
    y="Column",
    orientation="h",
    text="Missing Percent",
    title="<b>Missing Data by Column</b><br><span style='font-size: 12px; color: #6B7280;'>Percentage of null values per field</span>",
    template="plotly_white",
    height=380,
)

fig.update_traces(
    marker_color="#4F46E5",
    marker_line_width=0,
    texttemplate="%{text:.1f}%",
    textposition="outside",
    cliponaxis=False,
    hovertemplate="<b>%{y}</b><br>Missing: %{x:.1f}%<extra></extra>",
)

fig.update_layout(
    xaxis={
        "title": "Missing (%)",
        "range": [0, 110],
        "showgrid": True,
        "gridcolor": "#F3F4F6",
        "zeroline": False,
        "ticksuffix": "%",
    },
    yaxis={
        "title": "",
        "tickfont": {"size": 12, "color": "#374151"},
    },
    plot_bgcolor="white",
    paper_bgcolor="white",
    font={"family": "system-ui, -apple-system, sans-serif", "color": "#111827"},
    margin={"l": 20, "r": 40, "t": 60, "b": 40},
)

fig.show()

In [23]:
# Count duplicate rows and job identifiers.
print(f"Duplicate rows: {jobs.duplicated().sum():,}")
print(f"Duplicate Job IDs: {jobs['Job ID'].duplicated().sum():,}")
print(f"Unique Job IDs: {jobs['Job ID'].nunique():,}")


Duplicate rows: 10
Duplicate Job IDs: 1,134
Unique Job IDs: 1,228


In [24]:
# Select the fields needed for career-path analysis.
analysis_fields = [
    "Job ID",
    "Business Title",
    "Civil Service Title",
    "Job Category",
    "Career Level",
    "Job Description",
    "Minimum Qual Requirements",
    "Preferred Skills",
]

jobs[analysis_fields].head(10)


,Job ID,Business Title,Civil Service Title,Job Category,Career Level,Job Description,Minimum Qual Requirements,Preferred Skills
0,780346,"Public Health Inspector, Bureau of Food Safety...",PUBLIC HEALTH SANITARIAN,"Public Safety, Inspections, & Enforcement",Entry-Level,**OPEN TO CANDIDATE PERMANENT IN THE TITLE OR ...,1. A baccalaureate degree from an accredited c...,NaN
1,780851,Project Manager,ASSOCIATE PROJECT MANAGER,"Engineering, Architecture, & Planning",Experienced (non-manager),Hours: Full-Time – 35 Hours Work Location: 30-...,One year of full-time satisfactory experience ...,Candidates should have the ability to manage c...
2,781780,"Scientist (Water Ecology), I - Bureau of Envir...",SCIENTIST (WATER ECOLOGY),"Public Safety, Inspections, & Enforcement",Experienced (non-manager),The Office of Building Water System Oversight ...,1. A masters degree from an accredited college...,NaN
3,778992,"Associate Laboratory Microbiologist, Bureau of...",ASSOCIATE LABORATORY MICROBIOL,Health,Experienced (non-manager),The Division of Disease Control’s mission is t...,1. A baccalaureate degree from an accredited c...,NaN
4,772547,Executive Assistant,ADM MANAGER-NON-MGRL FRM M1/M2,"Technology, Data & Innovation",Experienced (non-manager),ONLY PERMANENT EMPLOYEES IN THE ADMINISTRATIVE...,1. A baccalaureate degree from an accredited c...,NaN
5,771080,"Plumber, Bureau of Operations",PLUMBER,Building Operations & Maintenance,Experienced (non-manager),OPEN TO PERMANENT EMPLOYEES IN THE CIVIL SERVI...,(1) Five years of full-time satisfactory exper...,NaN
6,782334,"ASSOCIATE RETIREMENT BENEFITS EXAMINER, LEVEL 1",ASSOCIATE RETIREMENT BENEFITS,"Finance, Accounting, & Procurement Policy, Res...",Entry-Level,The New York City Employees’ Retirement System...,1. An associate degree or completion of 60 sem...,"•\tStrong communication, both written and verb..."
7,759909,REGIONAL MANAGER,ADMINISTRATIVE JOB OPPORTUNITY,Social Services,Manager,"If you are hired provisionally in the title, y...",1. A baccalaureate degree from an accredited c...,• Exemplary Supervisory leadership and experie...
8,781758,Captain (Sludge Boat),CAPTAIN (SLUDGE BOAT),Building Operations & Maintenance,Experienced (non-manager),***IMPORTANT NOTE: Candidate(s) selected for t...,Qualification Requirements A valid Merchant Ma...,"Relative to current operations, it is specific..."
9,778100,Solutions Developer for the Office of HPD TECH,IT PROJECT SPECIALIST,"Technology, Data & Innovation",Experienced (non-manager),About the Agency: The New York City Departmen...,1. A baccalaureate degree from an accredited c...,NaN


In [25]:
# Plot the share of postings at each career level.
career_levels = (
    jobs["Career Level"]
    .fillna("Missing")
    .value_counts(normalize=True, dropna=False)
    .rename_axis("Career Level")
    .reset_index(name="Postings")
)

fig = px.bar(
    career_levels,
    x="Career Level",
    y="Postings",
    text="Postings",
    title="<b>Job Postings by Career Level</b>",
    template="plotly_white",
)

fig.update_traces(
    texttemplate="%{text:.1%}",
    textposition="outside",
    marker_color="#4F46E5",
    hovertemplate="<b>%{x}</b><br>Share: %{y:.1%}<extra></extra>",
)

fig.show()


In [26]:
# Plot the 15 job categories with the most postings.
top_categories = (
    jobs["Job Category"]
    .fillna("Missing")
    .value_counts()
    .head(15)
    .rename_axis("Job Category")
    .reset_index(name="Postings")
    .sort_values("Postings")
)

fig = px.bar(
    top_categories,
    x="Postings",
    y="Job Category",
    text="Postings",
    orientation="h",
    title="<b>Top 15 Job Categories</b>",
    template="plotly_white",
)  

fig.show()


In [27]:
# Measure missing, unique, and duplicate business titles.
title_quality = (
    jobs["Business Title"]
    .fillna("")
    .str.strip()
    .replace("", np.nan)
)

print(f"Missing or blank titles: {title_quality.isna().sum():,}")
print(f"Unique titles: {title_quality.nunique():,}")
print(f"Duplicate titles: {title_quality.duplicated().sum():,}")


Missing or blank titles: 0
Unique titles: 908
Duplicate titles: 1,454


In [30]:
# Summarize the lengths of titles, descriptions, and requirements.
text_quality = pd.DataFrame(
    {
        "title_characters": jobs["Business Title"].fillna("").str.len(),
        "description_words": jobs["Job Description"].fillna("").str.split().str.len(),
        "requirements_words": (
            jobs["Minimum Qual Requirements"].fillna("").str.split().str.len()
        ),
    }
)

text_quality.describe()


,title_characters,description_words,requirements_words
count,2362.000000,2362.000000,2362.000000
mean,33.160881,589.842930,150.583827
std,18.823157,260.528781,129.035985
min,7.000000,67.000000,0.000000
25%,19.000000,393.000000,71.000000
50%,27.000000,572.500000,137.000000
75%,42.000000,747.000000,202.000000
max,131.000000,1888.000000,1570.000000


In [31]:
# Summarize annual salary ranges and their midpoints.
annual_salaries = jobs.loc[
    jobs["Salary Frequency"].eq("Annual"),
    ["Salary Range From", "Salary Range To", "Career Level"],
].copy()

annual_salaries["salary_from"] = pd.to_numeric(
    annual_salaries["Salary Range From"], errors="coerce"
)

annual_salaries["salary_to"] = pd.to_numeric(
    annual_salaries["Salary Range To"], errors="coerce"
)

annual_salaries["salary_midpoint"] = (
    annual_salaries["salary_from"] + annual_salaries["salary_to"]
) / 2

annual_salaries[["salary_from", "salary_to", "salary_midpoint"]].describe()


,salary_from,salary_to,salary_midpoint
count,2173.000000,2173.000000,2173.000000
mean,77866.029287,102698.851247,90282.440267
std,33388.731352,43325.897870,35998.508486
min,0.000000,23243.000000,23243.000000
25%,60275.000000,72298.000000,66280.000000
50%,72066.000000,96305.000000,84107.000000
75%,87743.000000,118387.000000,104107.000000
max,385000.000000,415000.000000,400000.000000


In [32]:
# Plot the number of job postings published each month.
eda = jobs.copy()
eda["posting_date"] = pd.to_datetime(eda["Posting Date"], errors="coerce")

print(f"Invalid or missing posting dates: {eda['posting_date'].isna().sum():,}")

postings_by_month = (
    eda.dropna(subset=["posting_date"])
    .assign(month=lambda frame: frame["posting_date"].dt.to_period("M").astype(str))
    .groupby("month")
    .size()
    .reset_index(name="Postings")
)

fig = px.line(
    postings_by_month,
    x="month",
    y="Postings",
    markers=True,
    title="Job Postings by Month",
)

fig.show()


Invalid or missing posting dates: 0


The dataset has **2,362 rows** and **1,228 unique job IDs**. Most repeated IDs are internal and external versions of the same job. For example, job `540190` appears twice as `Assistant District Attorney - Economic Crimes Bureau`, once for each posting type. Before modeling, combine these records into one job and keep the posting types as metadata.

There are also **10 exact duplicate rows**. For example Job `776352`, `Associate Water Use Inspector`, contains the same internal record twice. These copies can be removed safely.


For the mvp model we can use `Business Title` as the embedding text since it is complete and relevant to our initial modeling.

We can keep `Civil Service Title` for clustering. For example, `Project Manager`, `Engineering Auditor`, and `Senior Project Administrator` all use the civil service title `ASSOCIATE PROJECT MANAGER`, which can help group related roles.

We can use `Job Category` to color and filter the map Some categories are broad. For example, `Executive Assistant` appears under `Technology, Data & Innovation`.

We can use `Career Level` to label seniority and direct possible career paths. The data is heavily skewed towards `Experienced(None-manager)` roles

- Experienced non-manager: **73.2%**
- Entry-level: **11.6%**
- Manager: **10.2%**
- Student: **2.8%**
- Executive: **2.2%**

The labels are useful but broad. For example, `Senior Project Manager` is still labeled `Experienced (non-manager)`. Keep seniority terms from the title when cleaning the text.

### Text fields for later experiments

`Job Description` is complete, and `Minimum Qual Requirements` is available for **98.5%** of rows. These fields can support a later model. Start with titles so repeated agency language does not affect the first set of clusters.

Keep `Preferred Skills` optional because it is missing from **50.8%** of rows. For example, the `Project Manager` posting includes construction-management skills, while the `Public Health Inspector` posting has no preferred-skills value.

### Preprocessing handoff

1. Remove the 10 exact duplicate rows.
2. Create one modeling row per `Job ID` and preserve all posting types.
3. Create `normalized_title` by fixing capitalization, tabs, repeated spaces, and minor punctuation differences.
4. Preserve terms such as `Junior`, `Senior`, `Lead`, `Manager`, `Director`, and `Level 1`.
5. Create `seniority_rank` from `Career Level`: `Student = 0`, `Entry-Level = 1`, `Experienced (non-manager) = 2`, `Manager = 3`, and `Executive = 4`.
6. Use `normalized_title` for the first embeddings. Keep the original title, civil service title, category, career level, and posting type for labels, filters, and validation.

For example:

```text
Assistant District Attorney -     Economic Crimes Bureau
-> assistant district attorney - economic crimes bureau
```

Leave salary out of the first embedding. The file mixes annual and hourly salaries, and some annual records have a minimum salary of zero. For example, `Systems Developer` has an annual range of `$0-$61,194`, while `Plumber, Bureau of Operations` has an hourly rate of `$64`. Salary needs separate validation before we use it as a career signal.

After preprocessing, validate the row count and required fields. Then generate the title embeddings and reduce them to three dimensions.